pam/integration-tests/ssh_test.go

```diff
@@ -0,0 +1,557 @@
+	server := httptest.NewServer(http.HandlerFunc(nil))
+	url, err := url.Parse(server.URL)
+	require.NoError(t, err, "Setup: Impossible to find a valid port to use")
+	sshdPort := url.Port()
+	server.Close()
+
+	sshd, sshdPidFile, sshdLogFile := sshdCommand(t, sshdPort, hostKey, forcedCommand, env, daemonize)
+	sshdStderr := safeBuffer{}
+	sshd.Stderr = &sshdStderr
+	if testing.Verbose() {
+		sshd.Stdout = os.Stdout
+		sshd.Stderr = os.Stderr
+	}
+
+	t.Log("Launching sshd with", sshd.Env, sshd.Args)
+	err = sshd.Start()
+	require.NoError(t, err, "Setup: Impossible to start sshd")
+
+	t.Cleanup(func() {
+		if testing.Verbose() || !t.Failed() {
+			return
+		}
+		sshdLog := filepath.Join(t.TempDir(), "sshd.log")
+		require.NoError(t, os.WriteFile(sshdLog, []byte(sshdStderr.String()), 0600),
+			"TearDown: Saving sshd log")
+		saveArtifactsForDebug(t, []string{sshdLog})
+	})
+
+	t.Cleanup(func() {
+		if sshd.Process == nil {
+			return
+		}
+
+		sshdExited := make(chan *os.ProcessState)
+		go func() {
+			processState, err := sshd.Process.Wait()
+			require.NoError(t, err, "TearDown: Waiting SSHd failed")
+			sshdExited <- processState
+		}()
+
+		t.Log("Waiting for sshd to be terminated")
+		select {
+		case <-time.After(sleepDuration(30 * time.Second)):
+			require.NoError(t, sshd.Process.Kill(), "TearDown: Killing SSHd failed")
+			if !testing.Verbose() {
+				t.Logf("SSHd stopped (killed)\n ##### STDERR #####\n %s \n ##### END #####",
+					sshdStderr.String())
+			}
+			t.Fatal("SSHd didn't finish in time!")
+		case state := <-sshdExited:
+			if !testing.Verbose() {
+				t.Logf("SSHd stopped (%s)\n ##### STDERR #####\n %s \n ##### END #####",
+					state, sshdStderr.String())
+			}
+			expectedExitCode := 255
+			if daemonize {
+				expectedExitCode = 0
+			}
+			require.Equal(t, expectedExitCode, state.ExitCode(), "TearDown: SSHd exited with %s", state)
+		}
+	})
+
+	if !daemonize {
+		// Sadly we can't wait for SSHd to be ready using net.Dial, since that will make sshd
+		// (when in debug mode) not to accept further connections from the actual test, but we
+		// can assume we're good.
+		t.Logf("SSHd started with pid %d and listening on port %s", sshd.Process.Pid, sshdPort)
+		return sshdPort
+	}
+
+	t.Cleanup(func() {
+		if !t.Failed() && !testutils.IsVerbose() {
+			return
+		}
+		contents, err := os.ReadFile(sshdLogFile)
+		require.NoError(t, err, "TearDown: Reading SSHd log failed")
+		t.Logf(" ##### LOG FILE #####\n %s \n ##### END #####", contents)
+	})
+
+	t.Cleanup(func() {
+		pidFileContent, err := os.ReadFile(sshdPidFile)
+		require.NoError(t, err, "TearDown: Reading SSHd pid file failed")
+		p := strings.TrimSpace(string(pidFileContent))
+		pid, err := strconv.Atoi(p)
+		require.NoError(t, err, "TearDown: Parsing SSHd pid file content: %q", p)
+		process, err := os.FindProcess(pid)
+		require.NoError(t, err, "TearDown: Finding SSHd process")
+		err = process.Kill()
```

Comment:
Nope, because when running as a daemon `sshd` has two processes, the parent and the forked one, and weirdly enough killing only the parent isn't enough to kill the forked child, so we need to take care of both or we'll have dangling sshd instances running around.

Reply:
Makes sense!